File: elixir/lib/2016/02_bathroom_security/aoc201602.ex
defmodule AOC2016.Day02 do
  @moduledoc """
  Advent of Code 2016, day 2: Bathroom Security
  """
  require AOC

  @doc """
  Parse input
  """
  def parse(puzzle_input) do
    puzzle_input |> String.split("\n")
  end

  @doc """
  Solve part 1
  """
  def part1(input) do
    input
    |> moves_to_keys(%{
      {-1, 1} => "1",
      {0, 1} => "2",
      {1, 1} => "3",
      {-1, 0} => "4",
      {0, 0} => "5",
      {1, 0} => "6",
      {-1, -1} => "7",
      {0, -1} => "8",
      {1, -1} => "9"
    })
    |> Enum.join()
  end

  @doc """
  Solve part 2
  """
  def part2(input) do
    input
    |> moves_to_keys(%{
      {2, 2} => "1",
      {1, 1} => "2",
      {2, 1} => "3",
      {3, 1} => "4",
      {0, 0} => "5",
      {1, 0} => "6",
      {2, 0} => "7",
      {3, 0} => "8",
      {4, 0} => "9",
      {1, -1} => "A",
      {2, -1} => "B",
      {3, -1} => "C",
      {2, -2} => "D"
    })
    |> Enum.join()
  end

  @doc """
  Convert a list of moves to a list of keys

  ## Examples:

      iex> moves_to_keys(["L", "RUD"], %{{-1, 0} => "1", {0, -1} => "2", {0, 0} => "3"})
      ["1", "2"]
  """
  def moves_to_keys(moves, keypad), do: moves_to_keys(moves, keypad, {0, 0}, [])
  def moves_to_keys([], _keypad, _pos, keys), do: Enum.reverse(keys)

  def moves_to_keys([move | moves], keypad, pos, keys) do
    {new_pos, key} = move |> String.to_charlist() |> make_moves(keypad, pos)
    moves_to_keys(moves, keypad, new_pos, [key | keys])
  end

  @doc """
  Make one sequence of moves

  ## Examples:

      iex> make_moves('UD', %{{-1, 0} => "1", {0, -1} => "2", {0, 0} => "3"}, {0, 0})
      {{0, -1}, "2"}
  """
  def make_moves([], keypad, pos), do: {pos, Map.get(keypad, pos)}

  def make_moves([move | moves], keypad, pos) do
    maybe_pos = make_move(move, pos)
    new_pos = if Map.has_key?(keypad, maybe_pos), do: maybe_pos, else: pos
    make_moves(moves, keypad, new_pos)
  end

  @doc """
  Make one move

  ## Examples:

      iex> make_move(?D, {2, 3})
      {2, 2}
      iex> make_move(?L, {0, 0})
      {-1, 0}
  """
  def make_move(?U, {x, y}), do: {x, y + 1}
  def make_move(?D, {x, y}), do: {x, y - 1}
  def make_move(?L, {x, y}), do: {x - 1, y}
  def make_move(?R, {x, y}), do: {x + 1, y}

  def main(args) do
    Enum.map(args, fn path -> AOC.solve(path, &parse/1, &part1/1, &part2/1) end)
  end
end

File: python/2016/02_bathroom_security/aoc201602.py
"""AoC 2, 2016: Bathroom Security."""

# Standard library imports
import pathlib
import sys


def parse_data(puzzle_input):
    """Parse input."""
    return [[parse_move(move) for move in moves] for moves in puzzle_input.split("\n")]


def parse_move(move):
    """Parse one move into coordinate deltas.

    ## Example:

    >>> parse_move("D")
    (0, -1)
    """
    return {"U": (0, 1), "D": (0, -1), "L": (-1, 0), "R": (1, 0)}[move]


def part1(data):
    """Solve part 1."""
    keys = (
        {(-1, 1): "1", (0, 1): "2", (1, 1): "3"}
        | {(-1, 0): "4", (0, 0): "5", (1, 0): "6"}
        | {(-1, -1): "7", (0, -1): "8", (1, -1): "9"}
    )
    return "".join(moves_to_keys(data, keys))


def part2(data):
    """Solve part 2."""
    keys = (
        {(2, 2): "1"}
        | {(1, 1): "2", (2, 1): "3", (3, 1): "4"}
        | {(0, 0): "5", (1, 0): "6", (2, 0): "7", (3, 0): "8", (4, 0): "9"}
        | {(1, -1): "A", (2, -1): "B", (3, -1): "C"}
        | {(2, -2): "D"}
    )
    return "".join(moves_to_keys(data, keys))


def moves_to_keys(moves, keys, pos_x=0, pos_y=0):
    """Convert a list of move sequences to key codes.

    ## Example:

    >>> keys = {(-1, 0): "1", (0, -1): "2", (0, 0): "3"}
    >>> list(moves_to_keys([[(-1, 0)], [(1, 0), (0, 1), (0, -1)]], keys))
    ['1', '2']
    """
    for move_sequence in moves:
        for dx, dy in move_sequence:
            if (pos_x + dx, pos_y + dy) in keys:
                pos_x += dx
                pos_y += dy
        yield keys[(pos_x, pos_y)]


def solve(puzzle_input):
    """Solve the puzzle for the given input."""
    data = parse_data(puzzle_input)
    yield part1(data)
    yield part2(data)


if __name__ == "__main__":
    for path in sys.argv[1:]:
        print(f"\n{path}:")
        solutions = solve(puzzle_input=pathlib.Path(path).read_text().strip())
        print("\n".join(str(solution) for solution in solutions))
